core/src/worker/workflow/managed_run.rs

```diff
@@ -0,0 +1,606 @@
+                    }
+                    RunActions::HeartbeatTimeout => me.heartbeat_timeout().map(|autocomplete| {
+                        if autocomplete {
+                            (
+                                Some(ActivationOrAuto::Autocomplete {
+                                    run_id: me.wfm.machines.run_id.clone(),
+                                }),
+                                None,
+                            )
+                        } else {
+                            (None, None)
+                        }
+                    }),
+                };
+                match res {
+                    Ok((maybe_act, maybe_wft_update)) => {
+                        me.send_update_response(maybe_act, maybe_wft_update);
+                    }
+                    Err(e) => {
+                        error!(error=?e, "Error in run machines");
+                        me.am_broken = true;
+                        me.update_tx
+                            .send(RunUpdateResponse {
+                                kind: RunUpdateResponseKind::Fail(FailRunUpdate {
+                                    run_id: me.wfm.machines.run_id.clone(),
+                                    err: e.source,
+                                    completion_resp: e.complete_resp_chan,
+                                }),
+                                span: Span::current(),
+                            })
+                            .expect("Machine can send update");
+                    }
+                }
+                (me, heartbeat_tx)
+            }
+            .instrument(span)
+        })
+        .await;
+    }
+
+    async fn incoming_wft(
+        &mut self,
+        wft: NewIncomingWFT,
+    ) -> Result<(Option<ActivationOrAuto>, RunUpdatedFromWft), RunUpdateErr> {
+        let activation = if let Some(h) = wft.history_update {
+            self.wfm.feed_history_from_server(h).await?
+        } else {
+            let r = self.wfm.get_next_activation().await?;
+            if r.jobs.is_empty() {
+                return Err(RunUpdateErr {
+                    source: WFMachinesError::Fatal(format!(
+                        "Machines created for {} with no jobs",
+                        self.wfm.machines.run_id
+                    )),
+                    complete_resp_chan: None,
+                });
+            }
+            r
+        };
+
+        if activation.jobs.is_empty() {
+            if self.wfm.machines.outstanding_local_activity_count() > 0 {
+                // If the activation has no jobs but there are outstanding LAs, we need to restart the
+                // WFT heartbeat.
+                if let Some(ref mut lawait) = self.waiting_on_la {
+                    if lawait.completion_dat.is_some() {
+                        panic!("Should not have completion dat when getting new wft & empty jobs")
+                    }
+                    lawait.heartbeat_timeout_task.abort();
```

Comment:
What happens to the result of this task when aborted? Do we care?

Reply:
The task has no result. We don't care